### PR TITLE
OpenAPI (redoc/swagger/spotlight elements) usage in environments that cant access CDN's

### PR DESCRIPTION
**Pull request by nielsvanhooy**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

i was working on migrating a project from fastapi to starlite and was missing this function
it is inspired on how FastAPI handles the offline usage for the OpenAPI pages.

specifally inspired of the following from FastAPI.

functions:
get_swagger_ui_html
get_redoc_html

google_font:
https://github.com/tiangolo/fastapi/pull/481

*full transparency. im a self learned "programmer" so if something is incorrect bear with me*

### CODE REVIEW

File: docs/usage/12-openapi/3-openapi-controller.md [hunk 1]
```diff
@@ -39,3 +39,35 @@ app = Starlite(
 
 See the [API Reference][starlite.openapi.controller.OpenAPIController] for full details on the `OpenAPIController` class
 and the kwargs it accepts.
+
+
+## OpenAPI usage in offline environments.
```
Comment (Goldziher) [inline docs/usage/12-openapi/3-openapi-controller.md:44]: I'd rewrite this thus

```suggestion
## CDN and offline file support

You can change the default download paths for JS and CSS bundles as well as google fonts by subclassing [OpenAPIController][starlite.openapi.controller.OpenAPIController]  and setting any of the following class variables:
```

File: starlite/openapi/controller.py [hunk 2]
```diff
@@ -44,6 +44,49 @@ class OpenAPIController(Controller):
     """
     URL to download a favicon from.
     """
+    redoc_google_fonts: bool = True
+    """
+    Google CDN Fonts in Redoc, use False for use in isolated environment
```
Comment (Goldziher) [inline starlite/openapi/controller.py:49]:
```suggestion
    Download google fonts via CDN. Should be set to `False` when not using a CDN. 
```

File: starlite/openapi/controller.py [hunk 3]
```diff
@@ -44,6 +44,49 @@ class OpenAPIController(Controller):
...
+    """
+    redoc_js_url: str = f"https://cdn.jsdelivr.net/npm/redoc@{redoc_version}/bundles/redoc.standalone.js"
+    """
+    Specify redoc_js_url for use in isolated environment
```
Comment (Goldziher) [inline starlite/openapi/controller.py:53]:
```suggestion
    Download url for the Redoc JS bundle. 
```

File: starlite/openapi/controller.py [hunk 4]
```diff
@@ -44,6 +44,49 @@ class OpenAPIController(Controller):
...
+    Default reverts to as used in redoc_version
```
Comment (Goldziher) [inline starlite/openapi/controller.py:54]:
```suggestion
```

**Review by Goldziher — COMMENTED** [review]
Looks good, please add tests for the default urls that have this structure: `f"https://cdn.jsdelivr.net/npm/redoc@{redoc_version}/bundles/redoc.standalone.js"`, I am not certain that the variable `redoc_version` will be populated correctly.

File: starlite/openapi/controller.py (hunk 2 above)
Comment (nielsvanhooy) [reply starlite/openapi/controller.py]: Changed to your suggestion. Thanks :)

File: docs/usage/12-openapi/3-openapi-controller.md (hunk 1 above)
Comment (nielsvanhooy) [reply docs/usage/12-openapi/3-openapi-controller.md]: Changed to your suggestion. Thanks :)

File: starlite/openapi/controller.py (hunk 3 above)
Comment (nielsvanhooy) [reply starlite/openapi/controller.py]: Changed to your suggestion. Thanks :)

File: starlite/openapi/controller.py (hunk 4 above)
Comment (nielsvanhooy) [reply starlite/openapi/controller.py]: Changed to your suggestion. Thanks :)
also changed it for the others.

**Comment by nielsvanhooy** [comment]
> Looks good, please add tests for the default urls that have this structure: `f"https://cdn.jsdelivr.net/npm/redoc@{redoc_version}/bundles/redoc.standalone.js"`, I am not certain that the variable `redoc_version` will be populated correctly.

if setting the variables for offline support, then it doesn't look at the default version and does not compose the CDN url.
in openapi/controller.py   i see that redoc_version defaults to "next", swagger ui to "4.14.0",  stoplight to "7.6.5"

i can add to the already existing tests an assert that verifies the default CDN urls if that is the way to go.

or do you want that the offline urls will compose based on the version specified?

that would require me to change some other things around.
let me know or otherwise we could discuss this further in Discord.

**Comment by Goldziher** [comment]
> > Looks good, please add tests for the default urls that have this structure: `f"https://cdn.jsdelivr.net/npm/redoc@{redoc_version}/bundles/redoc.standalone.js"`, I am not certain that the variable `redoc_version` will be populated correctly.
> 
> if setting the variables for offline support, then it doesn't look at the default version and does not compose the CDN url. in openapi/controller.py i see that redoc_version defaults to "next", swagger ui to "4.14.0", stoplight to "7.6.5"
> 
> i can add to the already existing tests an assert that verifies the default CDN urls if that is the way to go.
> 
> or do you want that the offline urls will compose based on the version specified?
> 
> that would require me to change some other things around. let me know or otherwise we could discuss this further in Discord.

simply a test to verify that the default cdn urls are composed correctly.

**Comment by nielsvanhooy** [comment]
> > > Looks good, please add tests for the default urls that have this structure: `f"https://cdn.jsdelivr.net/npm/redoc@{redoc_version}/bundles/redoc.standalone.js"`, I am not certain that the variable `redoc_version` will be populated correctly.
> > 
> > 
> > if setting the variables for offline support, then it doesn't look at the default version and does not compose the CDN url. in openapi/controller.py i see that redoc_version defaults to "next", swagger ui to "4.14.0", stoplight to "7.6.5"
> > i can add to the already existing tests an assert that verifies the default CDN urls if that is the way to go.
> > or do you want that the offline urls will compose based on the version specified?
> > that would require me to change some other things around. let me know or otherwise we could discuss this further in Discord.
> 
> simply a test to verify that the default cdn urls are composed correctly.

added test for the default CDN urls.
i had to do an:  assert client.app.openapi_config is not None
because of mypy checkings that would otherwise fail..

"Item "None" of "Optional[OpenAPIConfig]" has no attribute "openapi_controller"  [union-attr]"
i didn't found another way to proceed without this.
please advise if all is ok now :)

**Review by Goldziher — APPROVED** [review]
great, thanks a lot!

**Comment by Goldziher** [comment]
@all-contributors add @nielsvanhooy code

**Pull request merged**